go/base/static/js/src/apps/dialogue/layout.js

```diff
@@ -0,0 +1,143 @@
+(function(exports) {
+  var Grid = go.components.grid.Grid;
+
+
+  var DialogueStateLayout = Backbone.View.extend({
+    initialize: function(options) {
+      options = _.defaults(options, {numCols: 3});
+      this.numCols = options.numCols;
+      this.states = options.states;
+      this.setElement(this.states.view.$el);
+      this.grid = new Grid({numCols: this.numCols});
+
+      this.onDrag = this.onDrag.bind(this);
+      go.utils.bindEvents(this.bindings, this);
+    },
+
+    events: {
+      'dblclick': function(e) {
+        this.addAtEvent(e);
+      }
+    },
+
+    numCols: 3,
+
+    cellWidth: function() {
+      return this.$el.width() / this.numCols;
+    },
+
+    cellHeight: function(state) {
+      return state.$el.outerHeight(true);
+    },
+
+    colOffset: function(state) {
+      var center = (this.cellWidth(state) - state.$el.outerWidth(true)) / 2;
+      return center + 35;
+    },
+
+    rowMargin: function() {
+      return 30;
+    },
```

Comment:
The layout was initially using margins specified in css for figuring out cell widths, heights and offsets, but that actually makes things more complicated -- in order to retain the 3 column layout for existing dialogues, we'd need kind-of big margins, which complicates dragging states around, since the now how pretty big bounding boxes when colliding with the edges of the diagram, which makes moving states around annoying.